### PR TITLE
Split internal sealing from work preparation

### PR DESCRIPTION
**Pull request by keorn**
Previously the "miner" attempted to generate and import seal in `prepare_sealing` method. Now methods are separated out more and flow is based in `seals_internally` Engine method.

### CODE REVIEW

File: ethcore/src/miner/miner.rs [hunk 1]
```diff
@@ -411,12 +419,13 @@ impl Miner {
 			// | NOTE Code below requires transaction_queue and sealing_work locks.     |
 			// | Make sure to release the locks before calling that method.             |
 			// --------------------------------------------------------------------------
-			self.prepare_sealing(chain);
+			let (block, original_work_hash) = self.prepare_block(chain);
+			self.prepare_work(block, original_work_hash);
```
Comment (keorn) [inline ethcore/src/miner/miner.rs]: No splat/apply :(

Comment (rphmeier) [inline ethcore/src/miner/miner.rs]:
```rust
#![feature(fn_traits, unboxed_closures)]

use std::ops::Fn;

fn main() {
    Fn::call(&foo, (1, 2.0));    
}

fn foo(x: i32, y: f32) {
    println!("({}, {})", x, y);
}
```

(although having to account for the implicit self makes that impossible in this case)

Comment (keorn) [inline ethcore/src/miner/miner.rs]: Constant interaction with `Miner` struct discourages pure functions though.

File: ethcore/src/engines/mod.rs [hunk 2]
```diff
@@ -71,6 +71,8 @@ pub trait Engine : Sync + Send {
 	/// Block transformation functions, after the transactions.
 	fn on_close_block(&self, _block: &mut ExecutedBlock) {}
 
+	/// If true, generate_seal has to be implemented.
+	fn seals_internally(&self) -> bool { false }
```
Comment (rphmeier) [inline ethcore/src/engines/mod.rs]: nit: there should be a blank line following this

File: ethcore/src/miner/miner.rs [hunk 3]
```diff
@@ -804,7 +813,15 @@ impl MinerService for Miner {
...
-			self.prepare_sealing(chain);
+			trace!(target: "miner", "update_sealing: preparing a block");
+			let (block, original_work_hash) = self.prepare_block(chain);
+			if self.engine.seals_internally() {
+				trace!(target: "miner", "update_sealing: engine indicates internal sealing");
+				self.seal_and_import_block_internally(chain, block);
+			} else {
+				trace!(target: "miner", "update_sealing: engine does not seal internally, preparing work");
+				self.prepare_work(block, original_work_hash);
+			}
```
Comment (rphmeier) [inline ethcore/src/miner/miner.rs]: what's the difference between not sealing internally and not sealing at all?

could use a bit of documentation explaining the different kinds of sealing engines can do

Comment (keorn) [inline ethcore/src/miner/miner.rs]: Sure, will add. There is "internal sealing" which does not require work to be done and "work sealing": usual way.

File: ethcore/src/engines/mod.rs (hunk 2 above)
Comment (keorn) [inline ethcore/src/engines/mod.rs]: There is plenty of methods next to each other in trait definitions in this and other files in the codebase. Id argue that for definitions its fine, as it often groups method families.

File: ethcore/src/miner/miner.rs [hunk 4]
```diff
@@ -334,37 +331,49 @@ impl Miner {
...
+			})
+		} else {
+			trace!(target: "miner", "seal_block_internally: unable to generate seal internally");
+			Err(Some(block))
+		}
+	}
 
+	/// Uses Engine to seal the block internally and then imports it to chain.
+	fn seal_and_import_block_internally(&self, chain: &MiningBlockChainClient, block: ClosedBlock) -> bool {
 		if !block.transactions().is_empty() {
```
Comment (NikVolf) [inline ethcore/src/miner/miner.rs]: Tests triggering that if-path (or _not_ triggering) would be nice

File: ethcore/src/engines/mod.rs (hunk 2 above)
Comment (rphmeier) [inline ethcore/src/engines/mod.rs]: I haven't noticed it personally, in general it is typical Rust style to provide a line of blank space between each function definition, at least if there are docs. (see an example from `std`: [Iterator](https://doc.rust-lang.org/stable/src/core/up/src/libcore/iter/iterator.rs.html#38-2062) )

Comment (keorn) [inline ethcore/src/engines/mod.rs]: I see, but Engine had this sort of logical grouping already. @arkpar suggested to leave it.

File: ethcore/src/miner/miner.rs [hunk 5]
```diff
@@ -243,19 +245,15 @@ impl Miner {
 	}
 
 	/// Prepares new block for sealing including top transactions from queue.
-	#[cfg_attr(feature="dev", allow(match_same_arms))]
-	#[cfg_attr(feature="dev", allow(cyclomatic_complexity))]
```
Comment (arkpar) [inline ethcore/src/miner/miner.rs]: Any particular reason to remove these?

**Pull request merged**